integration-tests/tests/cases/mpc.rs

```diff
@@ -184,6 +185,179 @@ async fn test_basic_sign() {
+    );
+    assert!(
+        start.elapsed() >= Duration::from_secs(2),
+        "signature completed too quickly, expected a retry delay"
+    );
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_sign_request_retries_multiple_times() {
+    const NUM_NODES: usize = 3;
+    const TARGET_ATTEMPTS: usize = 10;
+
+    let mut signature_droppers = Vec::with_capacity(NUM_NODES);
+
+    let mut builder = MpcFixtureBuilder::default()
+        .only_generate_signatures()
+        .with_signature_timeout_ms(200);
+
+    for idx in 0..NUM_NODES {
+        let participant = Participant::from(idx as u32);
+        let (dropper, filter) = SignatureDropper::new(participant);
+        builder = builder.with_outgoing_message_filter(idx, filter);
+        signature_droppers.push(dropper);
+    }
+
+    let network = builder.build().await;
+
+    tokio::time::timeout(
+        Duration::from_millis(300),
+        network.wait_for_presignatures(2),
+    )
+    .await
+    .expect("should start with enough presignatures");
+
+    {
+        let mut log = network.output.msg_log.lock().await;
+        log.clear();
+    }
+
+    for dropper in &signature_droppers {
+        dropper.enable();
+    }
+
+    let request = sign_request(3);
+    for node in &network.nodes {
+        node.sign_tx.send(request.clone()).await.unwrap();
+    }
+
+    let required_messages = TARGET_ATTEMPTS * (network.nodes.len() - 1);
+
+    tokio::time::timeout(Duration::from_secs(10), async {
+        loop {
+            let signature_messages = {
+                let log = network.output.msg_log.lock().await;
+                log.iter()
+                    .filter(|entry| entry.starts_with("Signature"))
+                    .count()
+            };
+
+            if signature_messages >= required_messages {
+                break signature_messages;
+            }
+
+            tokio::time::sleep(Duration::from_millis(20)).await;
+        }
+    })
+    .await
+    .expect("expected at least ten signature retries before timeout");
+
+    for dropper in &signature_droppers {
+        dropper.disable();
+    }
+
+    tokio::time::sleep(Duration::from_millis(50)).await;
+
+    let dropped_messages: usize = signature_droppers
+        .iter()
+        .map(SignatureDropper::dropped)
+        .sum();
+
+    assert!(
+        dropped_messages > 0,
+        "expected to drop at least one signature message before retries succeeded"
+    );
+
+    let actions = tokio::time::timeout(Duration::from_secs(10), network.wait_for_actions(1))
+        .await
+        .expect("signature should eventually publish after retries");
+
+    assert!(
+        actions
+            .iter()
+            .any(|action| action.contains("RpcAction::Publish")),
+        "unexpected rpc actions: {actions:?}"
+    );
+
+    let signature_messages = {
+        let log = network.output.msg_log.lock().await;
+        log.iter()
+            .filter(|entry| entry.starts_with("Signature"))
+            .count()
```

Comment:
side note: Should we change how logs are represented? I just wanted some dumb way of keeping a log of what happened, without coupling the testing closely with internal implementation types. Curious to hear what you think after you have written your first tests with this approach.

Reply:
yeah I think we need more robust typing or better convenience helpers for this in the long term. It's going to be hard to follow what this is doing